From 74728a044d55aa09ab8137bfc00ebc1a86bdc7c3 Mon Sep 17 00:00:00 2001
From: ctang <ctang@cloudera.com>
Date: Thu, 22 Sep 2016 14:47:47 -0400
Subject: [PATCH 1066/1164] CDH-52840 HIVE-14774: Canceling query using Ctrl-C
 in beeline might lead to stale locks (Chaoyu
 Tang, reviewed by Jimmy Xiang, Mohit Sabharwal)

(cherry picked from commit 3237bb276b79d2a54d0a5d8a5ede400147017618)

Change-Id: I1a75ef06323b2e70608401ba3a303f6462d27a24
---
 .../hive/service/cli/operation/SQLOperation.java   |   18 +++++++++++-------
 1 file changed, 11 insertions(+), 7 deletions(-)

diff --git a/service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java b/service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java
index 767ecc7..34ddcdb 100644
--- a/service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java
+++ b/service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java
@@ -373,6 +373,17 @@ private void registerCurrentOperationLog() {
 
   private synchronized void cleanup(OperationState state) throws HiveSQLException {
     setState(state);
+
+    if (shouldRunAsync()) {
+      Future<?> backgroundHandle = getBackgroundHandle();
+      if (backgroundHandle != null) {
+        boolean success = backgroundHandle.cancel(true);
+        if (success) {
+          LOG.info("The running operation has been successfully interrupted.");
+        }
+      }
+    }
+
     if (driver != null) {
       driver.close();
       driver.destroy();
@@ -387,13 +398,6 @@ private synchronized void cleanup(OperationState state) throws HiveSQLException
       ss.deleteTmpErrOutputFile();
     }
 
-    if (shouldRunAsync()) {
-      Future<?> backgroundHandle = getBackgroundHandle();
-      if (backgroundHandle != null) {
-        backgroundHandle.cancel(true);
-      }
-    }
-
     // Shutdown the timeout thread if any, while closing this operation
     if ((timeoutExecutor != null) && (state != OperationState.TIMEDOUT) && (state.isTerminal())) {
       timeoutExecutor.shutdownNow();
-- 
1.7.9.5

